lib/location/UserLocation.ts

```diff
@@ -1,154 +1,99 @@
 
 /**
- * An interface around operations regarding the current user's location.
+ * A function type to track a user's location.
+ *
+ * @param accurracy The accurracy of which to track the user's location.
+ * @param callback Handles an {@link UserLocationTrackingUpdate}.
+ * @returns A function to stop tracking the user's location.
  */
-export interface UserLocation {
-  /**
-   * Requests foreground permission to track the user's location.
-   *
-   * @returns `true` if granted.
-   */
-  requestForegroundPermission: () => Promise<boolean>
-
-  /**
-   * Returns the user's last known location if authorized.
-   *
-   * While this method is faster than `currentLocation`, it may return
-   * an outdated location or even `null` since it doesn't attempt to
-   * query the user's current location.
-   */
-  lastKnownLocation: () => Promise<TrackedLocation | null>
-
-  /**
-   * Returns the user's current location if authorized.
-   */
-  currentLocation: () => Promise<TrackedLocation>
-
-  /**
-   * Invokes the given callback when the user's location changes.
-   *
-   * @param options
-   * @param callback a callback that accepts a `TrackedLocation` and runs whenever an update occurs
-   * @returns a function to stop tracking
-   */
-  track: (
-    callback: (location: TrackedLocation) => void,
-    options?: UserLocationTrackingOptions
-  ) => Promise<StopUserLocationTracking>
-}
+export type TrackUserLocation = (
+  accuracy: UserLocationTrackingAccurracy,
+  onUpdate: (update: UserLocationTrackingUpdate) => void
+) => StopUserLocationTracking
 
 /**
- * `UserLocation` operations implemented by expo.
+ * Tracks the user's location using expo.
+ *
+ * @param accurracy The accurracy of which to track the user's location.
+ * @param callback Handles an {@link UserLocationTrackingUpdate}.
+ * @returns A function to stop tracking the user's location.
  */
-export class ExpoUserLocation implements UserLocation {
-  async requestForegroundPermission () {
-    return (await ExpoLocation.requestForegroundPermissionsAsync()).granted
-  }
-
-  async lastKnownLocation () {
-    const location = await ExpoLocation.getLastKnownPositionAsync()
-    return location ? toTrackedLocation(location) : null
-  }
-
-  async currentLocation () {
-    return toTrackedLocation(await ExpoLocation.getCurrentPositionAsync())
-  }
-
-  async track (
-    callback: (location: TrackedLocation) => void,
-    options: UserLocationTrackingOptions = defaultTrackingOptions
-  ) {
-    const subscription = await ExpoLocation.watchPositionAsync(
-      {
-        accuracy: toExpoAccurracy(options.accuracy),
-        distanceInterval: options.minUpdateMetersDistance
-      },
-      (location) => callback(toTrackedLocation(location))
-    )
-    return subscription.remove as StopUserLocationTracking
-  }
-}
-
-const defaultTrackingOptions: UserLocationTrackingOptions = {
-  accuracy: UserLocationTrackingAccurracy.Medium
+export const expoTrackUserLocation = (
```

Comment:
It looks like this isn't intended to be used outside of this file and only called through the dependency system. Is there a way we can prevent accidental direct import/misuse of this function in the future? We could remove "export" but we'd need that for testing.

Reply:
Unfortunately ts doesn't have precise access control... Otherwise yeah I would agree.